### PR TITLE
Update addon links to work with new store

### PR DESCRIPTION
**Pull request by tmeasday**
This a draft because we need to consider the back-compat implications of needing to make these changes. Namely:

1. `global.__STORYBOOK_X` are not immediately available on boot, but instead only after `configure()` has been called. I could change this with some small refactorings.

2. `store.getSelection()` doesn't exist any more. This was an internal API, that makes a lot less sense now that the preview contains the selection. We could probably hack it so it works though 🤷 

3. Calling `store.raw()` isn't guaranteed to work in the v7 store (you'd need to pre-cache the store). Instead you should look at the story index, like I have done in this PR. I think this part is OK, although I wonder if we should abstract the access to the index a little bit.

WDYT @shilman?

### CODE REVIEW

File: addons/links/src/preview.ts [hunk 1]
```diff
@@ -36,51 +33,63 @@ const generateUrl = (id: string) => {
 const valueOrCall = (args: string[]) => (value: string | ((...args: string[]) => string)) =>
   typeof value === 'function' ? value(...args) : value;
 
+// Similar to `useStoryContext` from '@storybook/addons', except works outside a decorator/story
+// NOTE: this means it only tells us the 'global' rendered story, and will not do anything
+// sensible if more than one story is rendered to the screen (i.e. docs)
+function getStoryContext(): StoryContext {
+  // eslint-disable-next-line no-underscore-dangle
```
Comment (shilman) [inline addons/links/src/preview.ts:40]: disable for whole file since we use this in a few different places?

File: addons/links/src/preview.ts [hunk 2]
```diff
@@ -36,51 +33,63 @@ const generateUrl = (id: string) => {
...
-  const current = storyStore.fromId(storyId) || {};
-  const kindVal = resolver(idOrKindInput);
-  const storyVal = resolver(storyInput);
+  const current = getStoryContext();
+  const titleVal = resolver(idOrTitleInput);
+  const nameVal = resolver(nameInput);
 
-  const fromid = storyStore.fromId(kindVal);
+  const { stories } = getStoryIndex();
+  const fromid = stories[titleVal];
```
Comment (shilman) [inline addons/links/src/preview.ts:59]: This line doesn't make sense to me.
- `fromid` => `fromId`
- `titleVal` => `idOrTitleVal` ?

**Review by shilman — COMMENTED** [review]
Where would the `global.__STORYBOOK_X` refactor occur?

The other two comments make sense, esp the bit about not exposing the story index for now -- that should be an internal API until we figure out whether we want to expose it.

**Comment by tmeasday** [comment]
> Where would the global.__STORYBOOK_X refactor occur?

We would have to make the `WebPreview` and `StoryStore`'s constructors not expect any arguments, and instead pass them on the `initialize` call. It would be a bit of a pain, but could be done.

**Comment by shilman** [comment]
> What do you mean by internal API here?

I mean that addons shouldn't be accessing the story index durectly

**Comment by tmeasday** [comment]
> I mean that addons shouldn't be accessing the story index directly.

So are you saying the block of code inside the `find()` should be moved into the store or something?

**Comment by tmeasday** [comment]
Resolved 1&2 on `on-demand-store` by adding back-compat APIs.

Will address point 3 & probably update this addon to no use those APIs (this PR) in a later iteration.

**Comment by shilman** [comment]
Merged as part of https://github.com/storybookjs/storybook/pull/15871

**Pull request closed without merging**